### PR TITLE
[refactor](Nereids): new Date/Datetime parser to support more condition

### PR DESCRIPTION
**Pull request by jackwener**
## Proposed changes

- unify all Date/Datetime use one string-parser
- support microsecond & ZoneOffset both exist
- add many UT case
- add determineScale() to get scale of datetime, original code just get length of part after `.`
- reject more bad condition like `2022-01-01 00:00:00.`, we don't allow `.` without microsecond.
- .....

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java [hunk 1]
```diff
@@ -115,27 +96,72 @@ public DateLiteral(DateLiteral other, DataType type) {
...
     }
 
-    protected void init(String s) throws AnalysisException {
+    // replace 'T' with ' '
+    private static String replaceDelimiterT(String s) {
+        // Matcher matcher = Pattern.compile("^(\\d{2,4}-\\d{1,2}-\\d{1,2})T").matcher(s);
+        // if (matcher.find()) {
+        //     return matcher.group(1) + " " + s.substring(matcher.end());
+        // }
+        // return s;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java:105]: remove useless comment

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java [hunk 2]
```diff
@@ -115,27 +96,72 @@ public DateLiteral(DateLiteral other, DataType type) {
...
+                throw new AnalysisException("datetime literal [" + originalString + "] is invalid");
             }
-            year = DateUtils.getOrDefault(dateTime, ChronoField.YEAR);
-            month = DateUtils.getOrDefault(dateTime, ChronoField.MONTH_OF_YEAR);
-            day = DateUtils.getOrDefault(dateTime, ChronoField.DAY_OF_MONTH);
+
+            return dateTime;
         } catch (Exception ex) {
-            throw new AnalysisException("date literal [" + s + "] is invalid");
+            throw new AnalysisException("datetime literal [" + originalString + "] is invalid");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java:153]: why change to datetime? this is dateliteral

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java (hunk 1 above)
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java:105]: Will be done in next PR

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java (hunk 2 above)
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateLiteral.java:153]: Will be done in next PR.

**Review by englefly — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**